policy/src/main/java/org/apache/brooklyn/policy/ha/ElectPrimaryEffector.java

```diff
@@ -0,0 +1,440 @@
+                    ServiceProblemsLogic.updateProblemsIndicator(entity(), "primary", "No primary could be found");
+                    entity().sensors().set(Sensors.newSensor(Entity.class, params.get(PRIMARY_SENSOR_NAME)), null);
+                    entity().sensors().set(Attributes.SERVICE_UP, false);
+                    if (Lifecycle.RUNNING.equals( entity().getAttribute(Attributes.SERVICE_STATE_EXPECTED) )) {
+                        entity().sensors().set(Attributes.SERVICE_STATE_ACTUAL, Lifecycle.ON_FIRE);
+                    } else {
+                        entity().sensors().set(Attributes.SERVICE_STATE_ACTUAL, Lifecycle.STOPPED);
+                    }
+                    DynamicTasks.queue(Tasks.create("demote "+currentActive, new Demote(params)) ).getUnchecked();                
+                    return MutableMap.of("code", ResultCode.NO_PRIMARY_AVAILABLE, "message", "No primary available", "primary", null);
+                }
+                
+                if (newPrimary.equals(currentActive)) {
+                    // If there is a primary and it is unchanged, the effector will end.
+                    return MutableMap.of("code", ResultCode.PRIMARY_UNCHANGED, "message", "No change required", "primary", newPrimary);
+                }
+                
+                log.info("Detected new primary "+newPrimary+" at "+entity()+" (previously had "+currentActive+")");
+//                If a new primary is detected, the effector will:
+//                    * set the local entity to the STARTING state
+//                    * clear any "primary-election" problem
+//                    * publish the new primary in a sensor called `primary` (or the sensor set in `primary.sensor.name`)
+//                    * cancel any other ongoing promote calls, and if there is an ongoing demote call on the entity being promoted, cancel that also
+//                    * in parallel
+//                        * invoke `promote` (or the effector called `primary.promote.effector.name`) on the local entity or the entity being promoted
+//                        * invoke `demote` (or the effector called `primary.promote.effector.name`) on the local entity or the entity being demoted, if an entity is being demoted
+//                    * set service up true
+//                    * set the local entity to the RUNNING state
+                
+                boolean wasRunning = entity().sensors().get(Attributes.SERVICE_STATE_ACTUAL) == Lifecycle.RUNNING;
+                if (wasRunning) {
+                    log.debug("Transititioning "+entity()+" to starting while promoting/demoting");
+                    ServiceStateLogic.setExpectedState(entity(), Lifecycle.STARTING);
+                }
+                ServiceProblemsLogic.clearProblemsIndicator(entity(), "primary");
+                entity().sensors().set(Sensors.newSensor(Entity.class, params.get(PRIMARY_SENSOR_NAME)), newPrimary);
```

Comment:
my thinking was that `{state: STARTING}` (or `ON_FIRE`) indicates that it may be transitioning, and that the `primary` should not be relied on to be ready (just like we wouldn't rely on a `main.uri` until service is up).

by setting the primary we ensure it will get a `demote` call if another is selected.  and there maybe be other reasons it's useful to know what the `primary` should be even if promotion failed.  explicit transition would record that but feels redundant.

the ideal thing would be to be able to rerun failed tasks:  what we don't have is a way to rerun promotion because calling `electPrimary` a second time will be no-op.  and/or we could have an extra flag `repromoteIfUnchanged` on the effector if needed, or a `clearPrimary(andMaybeDisablePolicy)` effector.  but could be enhancements.